### PR TITLE
[Preview] Integrated cache: Adds ShardKey to DedicatedGatewayRequestOptions

### PR DESCRIPTION
**Pull request by jiajunpeng-msft**
# Pull Request Template

## Description

Currently, integrated cache is not distributed. Requests will randomly land on any of the Dedicated Gateway instances, which causes inconsistency.

We are introducing a new "RequestOption" called "ShardKey" to specify the Dedicated Gateway request shard key. Shard key will be used to route the request to the correlated instance.

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs [hunk 1]
```diff
@@ -317,6 +317,16 @@ private static async Task<Stream> BufferContentIfAvailableAsync(HttpResponseMess
                             requestMessage.Headers.TryAddWithoutValidation(key, request.Headers[key]);
                         }
                     }
+
+                    if (key.Equals(HttpConstants.HttpHeaders.DedicatedGatewayShardKey, StringComparison.OrdinalIgnoreCase) && request.Headers[key] != null)
+                    {
+                        string useHttp2String = Environment.GetEnvironmentVariable("UseHttp2");
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs:323]: you can use this to read environment variable https://github.com/Azure/azure-cosmos-dotnet-v3/blob/5ec7b4bcd7e2e0698736cd374f2196682227c913/Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs#L18

Comment (jiajunpeng-msft) [reply Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs]: Thanks for sharing this! Changed to use `ConfigurationManager`.

**Review by sourabh1007 — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs [hunk 2]
```diff
@@ -57,6 +57,42 @@ public class DedicatedGatewayRequestOptions
...
+        /// 
+        /// // For QueryRequestOptions
+        /// QueryRequestOptions requestOptions = new QueryRequestOptions
+        /// {
+        ///     DedicatedGatewayRequestOptions = dedicatedGatewayRequestOptions
+        /// };
+        /// ]]>
+        /// </code>
+        /// </example>
+#if PREVIEW
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:89]: When is backend ready for public preview? 
Ideally this needs to be public after-wards. 

Private preview: As discussed before public contract is not a requirement and can be worked around through using custom-headers right? 

/cc: @jcocchi

**Review by kirankumarkolli — CHANGES_REQUESTED** [review]
Preview contracts are for public previews. Waiting for clarifiation

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs (hunk 2 above)
Comment (jcocchi) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:89]: Sharding public preview is several months out and our private preview will be limited, at least in the initial phase. Using custom headers makes sense for now so we don't confuse existing customers with new options that aren't widely available yet

Comment (jiajunpeng-msft) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs]: This makes sense to me. I removed ShardKey request option and modified the http2 to be used when it's Dedicated Gateway request and has the environment variable set.

Let me know if this makes sense to you @kirankumarkolli @jcocchi.

**Comment by kirankumarkolli** [comment]
Closing inactive PRs older than 3-months, please re-active it needed.

**Pull request closed without merging**